### ISSUE TITLE
MSVC question

### ISSUE THREAD

**Issue by thomassligon** [AMICI-dev/AMICI#1264]
@dweindl Hello Daniel, I have a question about support for MSVC. After I tested the scripts (`installOpenBLAS.ps1` and `compileBLAS.cmd`) locally and submitted them to Git, you made some improvements and everything was working fine on Travis, and we finished by updating the installation documentation. Now, in order to run AMICI, I need to install `openBLAS` locally, but I am having some difficulty. The `cmake` fails and terminates with this error:
`utest\openblas_utest.exe : fatal error LNK1120: 14 unresolved externals`
Did you have anything like that on Travis?
Do you know if the latest Travis log is still available for comparison purposes?
[out-2020-09-07.txt](https://github.com/AMICI-dev/AMICI/files/5184032/out-2020-09-07.txt)

**Comment by dweindl** [comment]
Hi Tom, 

you can always find the current (and old) build status at https://travis-ci.com/github/AMICI-dev/AMICI/builds

I can see there, that linking `utest\openblas_utest.exe` also fails there. Did not notice that last time.

However, `utest\openblas_utest.exe` sounds like OpenBLAS unittests. If they fail to link, that is not necessarily a problem for AMICI. As you can see in the travis results, AMICI still works as expected.

**Comment by thomassligon** [comment]
Thanks, that gets me closer. Now I have discovered that I was getting the old import error, so I decided to run `pip install amici`, but that fails with
```
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Iamici/include -Iamici/ThirdParty/SuiteSparse/KLU/Include/ -Iamici/ThirdParty/SuiteSparse/AMD/Include/ -Iamici/ThirdParty/SuiteSparse/COLAMD/Include/ -Iamici/ThirdParty/SuiteSparse/BTF/Include/ -Iamici/ThirdParty/SuiteSparse/SuiteSparse_config -Iamici/ThirdParty/SuiteSparse/include -Iamici/ThirdParty/sundials/include -Iamici/ThirdParty/sundials/src -Iamici/ThirdParty/gsl/ -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\ATLMFC\include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include -IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt /EHsc /Tpamici\src\cblas.cpp /Fobuild\temp.win-amd64-3.7\amici\src\cblas.obj -IC:BLASOpenBLAS-v0.3.10OpenBLAS-0.3.10 /std:c++14
    cblas.cpp
    amici\src\cblas.cpp(16): fatal error C1083: Cannot open include file: 'cblas.h': No such file or directory
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.27.29110\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2

```
The bad part is
`-IC:BLASOpenBLAS-v0.3.10OpenBLAS-0.3.10`
where the slashes are missing, but, in travis.yml, we have
`- export BLAS_LIBS='/LIBPATH:C:\\BLAS\\lib openblas.lib' BLAS_CFLAGS='-IC:\\BLAS\\OpenBLAS-v0.3.10\\OpenBLAS-0.3.10'`
which looks good.
In addition, I found two versions of `cblas.h`, one in 
`C:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10`
and one in
`C:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10\generated`
and that is presumably what we need.
Just as a sidelight, I wanted to save the output of `pip install amici`, and used a command line
`pip install amici 2>&1 3>&1 1>out-2020-09-08a.txt`
but the file `out-2020-09-08a.txt` is very small and doesn't contain any of the error messages from `cl.exe` etc. It looks like `pip install amici` is putting them somewhere other than `stderr` or `stdout`.

**Comment by dweindl** [comment]
> The bad part is
> `-IC:BLASOpenBLAS-v0.3.10OpenBLAS-0.3.10`
> where the slashes are missing

Should be easy to fix then. 

> In addition, I found two versions of `cblas.h`, one in
> `C:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10`
> and one in
> `C:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10\generated`
> and that is presumably what we need.

I'd assume that either of them will do the job, but feel free to compare the contents.

> Just as a sidelight, I wanted to save the output of `pip install amici`, and used a command line
> `pip install amici 2>&1 3>&1 1>out-2020-09-08a.txt`
> but the file `out-2020-09-08a.txt` is very small and doesn't contain any of the error messages from `cl.exe` etc. It looks like `pip install amici` is putting them somewhere other than `stderr` or `stdout`.

Redirecting output is not really an AMICI issue. I am not sure where else the output would be written to if neither to stdout nor to stderr. So I'd think this is rather an issue of your redirection. See e.g. here https://support.microsoft.com/en-us/help/110930/redirecting-error-messages-from-command-prompt-stderr-stdout - the order may matter (I'm not a Windows user).

**Comment by thomassligon** [comment]
I am closing this because the question has been answered, but there is still some work left to do.

The issue with the double backslashes is actually a bit interesting. In Travis, the export statement that creates the environment variables runs in Git BASH, which treats the backslash as an escape character, so it needs to be double. In Windows, the environment variables can be created in many ways, including the PowerShell script and the result is a correct Windows path definition. But it looks like python setuptools does some string manipulation that also treats the backslash as an escape character, so we need the double backslash here as well. I hope it doesn't create a conflict somewhere else. It might be better to use the forward slash here.

I wanted to fix the link error in utest, and created an issue in OpenBLAS. As it turns out, the 
`-DBUILD_DOUBLE=1` statement causes `cmake` to create a version of OpenBLAS that only contains double precision. That is not supported, but will probably work anyway. The unresolved external symbols in linking utest are all "non-double" routines, which explains the error.

The issue I opened in OpenBLAS is [https://github.com/xianyi/OpenBLAS/issues/2830](https://github.com/xianyi/OpenBLAS/issues/2830) and has been closed and replaced by 
[Make building the tests for individual data types conditional on the respective BUILD option #2833](https://github.com/xianyi/OpenBLAS/pull/2833) which should be fixed in the next version (0.3.11).
